tests/unit/s2n_sslv3_test.c

```diff
@@ -0,0 +1,139 @@
+/*
+* Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+*
+* Licensed under the Apache License, Version 2.0 (the "License").
+* You may not use this file except in compliance with the License.
+* A copy of the License is located at
+*
+*  http://aws.amazon.com/apache2.0
+*
+* or in the "license" file accompanying this file. This file is distributed
+* on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+* express or implied. See the License for the specific language governing
+* permissions and limitations under the License.
+*/
+
+#include "s2n_test.h"
+#include "testlib/s2n_testlib.h"
+#include "utils/s2n_random.h"
+
+#define S2N_TEST_DATA_SIZE 100
+
+S2N_RESULT s2n_test_send_receive_data(struct s2n_connection *sender, struct s2n_connection *receiver)
+{
+    uint8_t test_data[S2N_TEST_DATA_SIZE] = { 0 };
+    struct s2n_blob test_data_blob = { 0 };
+    EXPECT_SUCCESS(s2n_blob_init(&test_data_blob, test_data, sizeof(test_data)));
+    EXPECT_OK(s2n_get_public_random_data(&test_data_blob));
+
+    /* Send data */
+    s2n_blocked_status blocked = S2N_NOT_BLOCKED;
+    ssize_t bytes_written = 0;
+    while (bytes_written < S2N_TEST_DATA_SIZE) {
+        ssize_t w = s2n_send(sender, test_data + bytes_written, S2N_TEST_DATA_SIZE - bytes_written, &blocked);
+        EXPECT_TRUE(w >= 0);
+        bytes_written += w;
+    }
+
+    /* Receive data */
+    uint8_t buffer[S2N_TEST_DATA_SIZE] = { 0 };
+    ssize_t bytes_received = 0;
+    while (bytes_received < S2N_TEST_DATA_SIZE) {
+        ssize_t r = s2n_recv(receiver, buffer + bytes_received, S2N_TEST_DATA_SIZE - bytes_received, &blocked);
+        EXPECT_TRUE(r > 0);
+        bytes_received += r;
+    }
+
+    EXPECT_BYTEARRAY_EQUAL(test_data, buffer, S2N_TEST_DATA_SIZE);
+
+    return S2N_RESULT_OK;
+}
+
+int main(int argc, char **argv)
+{
+    BEGIN_TEST();
+
+    if (!s2n_hmac_is_available(S2N_HMAC_SSLv3_MD5)) {
+        /* AWS-LC should support SSLv3. */
+        EXPECT_FALSE(s2n_libcrypto_is_awslc());
+
+        /* Other libcryptos may not support SSLv3, so the tests are skipped. */
+        END_TEST();
+    }
+
+    DEFER_CLEANUP(struct s2n_cert_chain_and_key *rsa_chain_and_key = NULL, s2n_cert_chain_and_key_ptr_free);
+    EXPECT_SUCCESS(s2n_test_cert_chain_and_key_new(&rsa_chain_and_key,
+            S2N_DEFAULT_TEST_CERT_CHAIN, S2N_DEFAULT_TEST_PRIVATE_KEY));
+
+    DEFER_CLEANUP(struct s2n_cert_chain_and_key *ecdsa_chain_and_key = NULL, s2n_cert_chain_and_key_ptr_free);
+    EXPECT_SUCCESS(s2n_test_cert_chain_and_key_new(&ecdsa_chain_and_key,
+            S2N_DEFAULT_ECDSA_TEST_CERT_CHAIN, S2N_DEFAULT_ECDSA_TEST_PRIVATE_KEY));
+
+    char dhparams_pem[S2N_MAX_TEST_PEM_SIZE] = { 0 };
+    EXPECT_SUCCESS(s2n_read_test_pem(S2N_DEFAULT_TEST_DHPARAMS, dhparams_pem, S2N_MAX_TEST_PEM_SIZE));
```

Comment:
Does SSLv3 require DH params? I'm curious because I'm writing an SSLv3 integ tests right now and I didn't configure any DH params to get the tests working.

Reply:
Apparently s2nd always has some dhparams that it loads into the config. So that's probably why.

Reply:
No, it doesn't require DHE, but it supports it. This test loops over all SSLv3 cipher suites, and some of them are DHE. For example:
https://github.com/aws/s2n-tls/blob/8ee7a3cb07dbd5020310023ad12e2acb151f1b09/tls/s2n_cipher_suites.c#L261-L273

Maybe the happy path integ test doesn't test any DHE cipher suites 🤔

Reply:
> Actually s2nd always has some dhparams that it loads into the config. So that's probably why.

Ah ok - that makes sense!